[Analytics Hub] Exit Analytics Hub and display an error if date range selection fails

rachelmcr opened the pull request:
Part of: #8213

## Description

When selecting a date range fails, we now pop out of the Analytics Hub back to the My Store view and display a notice about the error.

We expect this should only happen in rare cases, when `AnalyticsHubTimeRangeSelection` fails to unwrap the current and previous time ranges.

## Changes

* In the analytics hub view model, we now catch `AnalyticsHubTimeRangeSelection.TimeRangeGeneratorError` errors. We set an error notice and set `errorSelectingTimeRange` to true.
* In the analytics hub view, we use the system notice presenter to enqueue the error notice when the hosting controller disappears, and pop the hosting controller when `errorSelectingTimeRange` is true.

## Testing

You can force a failed date range selection by temporarily editing one of the methods that unwraps the time range to always throw an error:

https://github.com/woocommerce/woocommerce-ios/blob/0af536a4dacda18538078cb996a298a98afde553/WooCommerce/Classes/ViewRelated/Dashboard/Analytics%20Hub/Time%20Range/AnalyticsHubTimeRangeSelection.swift#L68-L73

Then:

1. Build and run the app.
2. Tap the "See more" button.
3. Confirm that you are popped out of the Analytics Hub and an error notice appears on the My Store screen.

## Screenshots

https://user-images.githubusercontent.com/8658164/205310974-8721ae3c-86b3-4f08-81c0-d895141b6816.mp4




## Submitter Checklist

Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

rachelmcr committed:
Display a notice when date range selection fails

rachelmcr committed:
When date range selection fails, pop the view and display the notice on the dashboard

ealeksandrov commented on WooCommerce/Classes/ViewRelated/Dashboard/Analytics Hub/AnalyticsHubView.swift:
```diff
@@ -5,15 +5,33 @@ import SwiftUI
+        super.viewWillDisappear(animated)
+
+        // Show any notice that should have been presented before the underlying disappears.
+        enqueuePendingNotice(rootView.viewModel.notice, using: systemNoticePresenter)
```
Xcode shows a warning here for each navigation:

```
Accessing StateObject's object without being installed on a View. This will create a new instance each time.
```

ealeksandrov reviewed (commented):
It works well in my testing!
Just a few small considerations/ideas for improvement:

1. In the `trunk` we have additional refresh action which error may be unhandled.
2. Do you think we can connect dismiss + notice actions directly, so it's more obvious without relying on UIKit lifecycle (viewWillDisappear)? For example: `dismiss(with notice: Notice?)` on HostingVC can trigger both actions with clear intent.
3. `errorSelectingTimeRange` feels like duplicated information when we already have notice set and error thrown. Can we trigger notice+dismiss in the `catch` clause?

rachelmcr committed:
Merge branch 'trunk' into issue/8213-date-range-error-handling

rachelmcr committed:
Link the dismiss and notice actions more directly

rachelmcr committed:
Dismiss Analytics view when a notice is set

rachelmcr commented:
Thanks for the review @ealeksandrov! I took a look at your suggestions; let me know what you think.

> In the trunk we have additional refresh action which error may be unhandled.

Thanks for the heads up! I merged `trunk` to take a look at the refresh action and handled the error with the changes outlined below.

> Do you think we can connect dismiss + notice actions directly, so it's more obvious without relying on UIKit lifecycle (viewWillDisappear)? For example: dismiss(with notice: Notice?) on HostingVC can trigger both actions with clear intent.

I like the idea of making the intent clearer like this, but if we don't rely on the UIKit lifecycle it will change the timing of the notice. Using `viewWillDisappear` ensures that the notice doesn't appear before the view is dismissed. Otherwise, the notice shows up while the view is still visible:

https://user-images.githubusercontent.com/8658164/205677047-af666640-3059-4ba0-97b9-9e48d8c3be48.mp4

I prefer the timing that it has now, but I've updated the dismiss action in 7aa749f so it's more clearly connected to the notice. (This also resolves the Xcode warning you observed.) This way it still uses the UIKit lifecycle for the notice timing, but it also makes it clearer that we're both dismissing the view and handling the notice.

> errorSelectingTimeRange feels like duplicated information when we already have notice set and error thrown. Can we trigger notice+dismiss in the catch clause?

That's reasonable. It feels a little odd to me to call UI code directly in the view model, so I've updated the catch clause in 228f144 to just set the notice (removing `errorSelectingTimeRange`). Then, the view handles the dismiss action with the provided notice when the notice is set. (I've moved that to a separate `.onReceive` modifier to handle both the initial sync and refresh actions.)

ealeksandrov reviewed (approved):
Thanks for all changes and the demo, I agree that current implementation looks better!
One last nit: can we rename `notice` property and its comment to describe that setting it will dismiss the view? Something like "dismissNotice"?

rachelmcr committed:
Rename Notice property and update documentation
